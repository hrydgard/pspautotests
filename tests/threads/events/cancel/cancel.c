#include "../sub_shared.h"

SETUP_SCHED_TEST;

#define CANCEL_TEST(title, flag, pattern) { \
	int result = sceKernelCancelEventFlag(flag, pattern, NULL); \
	if (result == 0) { \
		printf("%s: OK\n", title); \
	} else { \
		printf("%s: Failed (%X)\n", title, result); \
	} \
	PRINT_FLAG(flag); \
}

#define CANCEL_TEST_WITH_WAIT(title, flag, pattern) { \
	int waitThreads = 99; \
	int result = sceKernelCancelEventFlag(flag, pattern, &waitThreads); \
	if (result == 0) { \
		printf("%s: OK (%d waiting)\n", title, waitThreads); \
	} else { \
		printf("%s: Failed (%X, %d waiting)\n", title, result, waitThreads); \
	} \
	PRINT_FLAG(flag); \
}

void foo() {
}

int main(int argc, char **argv) {
	SceUID flag = sceKernelCreateEventFlag("cancel", 0, 0, NULL);
	SceKernelEventFlagInfo flaginfo;

	// Something weird is happening: first time status is garbled?
	sceKernelReferEventFlagStatus(flag, &flaginfo);

	CANCEL_TEST("To all set", flag, 0xFFFFFFFF);
	CANCEL_TEST("To all clear", flag, 0x00000000);
	CANCEL_TEST("Some pattern", flag, 0xABCDEF45);

	CANCEL_TEST_WITH_WAIT("To all set", flag, 0xFFFFFFFF);
	CANCEL_TEST_WITH_WAIT("To all clear", flag, 0x00000000);
	CANCEL_TEST_WITH_WAIT("Some pattern", flag, 0xABCDEF45);

	sceKernelDeleteEventFlag(flag);

	BASIC_SCHED_TEST("NULL",
		result = sceKernelCancelEventFlag(0, 0x00000000, NULL);
	);
	BASIC_SCHED_TEST("Other unset",
		result = sceKernelCancelEventFlag(flag2, 0x00000000, NULL);
	);
	BASIC_SCHED_TEST("Other set",
		result = sceKernelCancelEventFlag(flag2, 0xFFFFFFFF, NULL);
	);
	BASIC_SCHED_TEST("Same unset",
		result = sceKernelCancelEventFlag(flag1, 0x00000000, NULL);
	);
	BASIC_SCHED_TEST("Same set",
		result = sceKernelCancelEventFlag(flag1, 0xFFFFFFFF, NULL);
	);
	BASIC_SCHED_TEST("Cancel other with threads",
		int n;
		result = sceKernelCancelEventFlag(flag2, 0xFFFFFFFF, &n);
	);
	BASIC_SCHED_TEST("Cancel waited with threads",
		int n;
		result = sceKernelCancelEventFlag(flag1, 0xFFFFFFFF, &n);
	);

	CANCEL_TEST("NULL", 0, 0);
	CANCEL_TEST("Invalid", 0xDEADBEEF, 0);
	CANCEL_TEST("Deleted", flag, 0);

	return 0;
}