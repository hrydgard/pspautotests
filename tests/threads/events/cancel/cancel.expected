To all set: OK
Event flag: OK (size=52,name='cancel',attr=0,init=00000000,cur=FFFFFFFF,wait=0)
To all clear: OK
Event flag: OK (size=52,name='cancel',attr=0,init=00000000,cur=00000000,wait=0)
Some pattern: OK
Event flag: OK (size=52,name='cancel',attr=0,init=00000000,cur=ABCDEF45,wait=0)
To all set: OK (0 waiting)
Event flag: OK (size=52,name='cancel',attr=0,init=00000000,cur=FFFFFFFF,wait=0)
To all clear: OK (0 waiting)
Event flag: OK (size=52,name='cancel',attr=0,init=00000000,cur=00000000,wait=0)
Some pattern: OK (0 waiting)
Event flag: OK (size=52,name='cancel',attr=0,init=00000000,cur=ABCDEF45,wait=0)
NULL: A1B1C2E1D1F2 (thread=800201B5, thread bits=00000000, main=8002019A)
Other unset: A1B1C2E1D1F2 (thread=800201B5, thread bits=00000000, main=00000000)
Other set: A1B1C2E1D1F2 (thread=800201B5, thread bits=00000000, main=00000000)
Same unset: A1B1C2D1E2F1 (thread=800201A9, thread bits=00000000, main=00000000)
Same set: A1B1C2D1E2F1 (thread=800201A9, thread bits=FFFFFFFF, main=00000000)
Cancel other with threads: A1B1C2E1D1F2 (thread=800201B5, thread bits=00000000, main=00000000)
Cancel waited with threads: A1B1C2D1E2F1 (thread=800201A9, thread bits=FFFFFFFF, main=00000000)
NULL: Failed (8002019A)
Event flag: Failed (00000000)
Invalid: Failed (8002019A)
Event flag: Invalid (8002019A)
Deleted: Failed (8002019A)
Event flag: Invalid (8002019A)
